### ISSUE TITLE
Esbuild plugin does up inject/upload all source and map files on build

### ISSUE THREAD

**Issue by Fwang36** [getsentry/sentry-javascript-bundler-plugins#351]
### Is there an existing issue for this?

- [X] I have checked for existing issues https://github.com/getsentry/sentry-javascript/issues
- [X] I have reviewed the documentation https://docs.sentry.io/
- [X] I am using the latest SDK release https://github.com/getsentry/sentry-javascript/releases

### How do you use Sentry?

Sentry Saas (sentry.io)

### Which SDK are you using?

@sentry/browser

### SDK Version

@sentry/esbuild-plugin 2.4.0

### Framework Version

react 18.2

### Steps to Reproduce

1. Enable splitting in esbuild config with `splitting: true`
2. Create a dynamic import
3. Add the sentryEsbuildPlugin()
4. build

My esbuild config - 
![Screenshot 2023-07-19 at 10 06 24 AM](https://github.com/getsentry/sentry-javascript/assets/96670625/958dfe52-40ed-45fc-a2a2-1b4df3787e9b)

My output files - 
<img width="234" alt="Screenshot 2023-07-19 at 10 06 39 AM" src="https://github.com/getsentry/sentry-javascript/assets/96670625/49e7af64-efd3-4c37-9a1f-65b9e317b15d">

Upload logs where only app.js and app.js.map being uploaded - 
<img width="864" alt="Screenshot 2023-07-19 at 10 07 09 AM" src="https://github.com/getsentry/sentry-javascript/assets/96670625/5fb778af-e997-4898-924c-93a5b4352fb9">

### Expected Result

All the files outputted by the bundler should be injected and uploaded

### Actual Result

Not all files outputted by the bundler are being injected and uploaded.

**Comment by lforst** [comment]
Took a look at this. Unfortunately, due to the restrictive nature of the esbuild plugin API we will probably not get debug ID injection working in combination with `splitting: true`. In these cases we will probably need to use `options.uploadLegacySourcemaps`.

**Comment by lforst** [comment]
Putting in backlog and keeping an eye on the esbuild API. Maybe in the future this will be possible.

**Comment by muuvmuuv** [comment]
Hey, I don't know if this is related or not, but when using the new Angular with esbuild and inject sentry esbuild plugin I see source maps in `buildArtifactPaths` but after `glob(globAssets)` all source maps get lost. Is that because the artifacts are not yet written?

**Comment by lforst** [comment]
@muuvmuuv did you provide an `assets` option?

**Comment by muuvmuuv** [comment]
@lforst yeah, ofc. I also created my own plugin just to print out all build artifacts to proove my assets glob is correct but sadly does not work. Since Angular does not support plugins atm I hacked around my own "cli" but it works perfectly for anything else: https://hastebin.com/share/lawidivura.javascript

**Comment by muuvmuuv** [comment]
@lforst yeah, ofc. I also created my own plugin just to print out all build artifacts to proove my assets glob is correct but sadly does not work. Since Angular does not support plugins atm I hacked around my own "cli" but it works perfectly for anything else: https://hastebin.com/share/lawidivura.javascript

**Comment by muuvmuuv** [comment]
@lforst yeah, ofc. I also created my own plugin just to print out all build artifacts to proove my assets glob is correct but sadly does not work. Since Angular does not support plugins atm I hacked around my own "cli" but it works perfectly for anything else: https://hastebin.com/share/lawidivura.javascript

**Comment by lforst** [comment]
Can you share what you defined in your assets option? I think it might be wrong.

**Comment by muuvmuuv** [comment]
I tried `./**/*.js.map`, `*.js.map` and `./www/browser/*.js.map`

**Comment by lforst** [comment]
You could add debug: true - maybe the plugin will tell you what is wrong. Generally I believe you should set assets to `**/*.js.map` (without dot in front) and that should work.

**Comment by muuvmuuv** [comment]
Nope, this also does not work. I guess this has something to do with the files not beeing actually created but used virtual until Angular finished all post tasks so Sentry wont find the files via "glob" function. The path of every build artifact is /project/chunk.xxxx.js.map but this file only exist virtually.

```
⠸ Building...[sentry-esbuild-plugin] Warning: Didn't find any matching sources for debug ID upload. Please check the `sourcemaps.assets` option.
```

I created this plugin to see the build artifacts. Got it from here which sound related but idk https://github.com/floydspace/serverless-esbuild/issues/145#issuecomment-861749419:

```js
{
    name: 'plugin-sm',
    setup: (build) => {
      build.onEnd((args) => {
        for (const file of args.outputFiles) {
          if (/\.map$/.test(file.path)) {
            console.log(file)
          }
        }
      })
    },
  }
```

It prints:


```
{
  path: '/Users/marvin/Developer/PROJECT/chunk-IVCW2ILW.js.map',
  contents: <Buffer 7b 22 76 63 ... 3634 more bytes>,
  hash: 'LYLq5nfs65g',
  text: [Getter]
}
{
  path: '/Users/marvin/Developer/PROJECT/chunk-OZZSOK5O.js.map',
  contents: <Buffer 7b 22 76 6e 69 63 ... 1101 more bytes>,
  hash: 'noq8VdjXkkc',
  text: [Getter]
}
```

**Comment by lforst** [comment]
Well this is not an error it is more like a warning and I would suggest following what it says. Did you clean your output folder before running the plugin with the `**/*.js` option? In general, I recommend completely omitting the assets option since the plugin should find output files by itself.

**Comment by wSedlacek** [comment]
I am running into this issue too when using Angular and Nx.
It seems like the `glob` is running before all the files have been saved to disk and thus not finding them.
https://github.com/getsentry/sentry-javascript-bundler-plugins/blob/a097e3249129d30a058d10f41a66960f17ca1647/packages/bundler-plugin-core/src/index.ts#L494-L501

I have confirmed that checking my `outputFiles` in `onEnd` does have the source map files.
Perhaps there is a way to move the source map upload to after everything has been written?
Or maybe the `outputFiles` could be used directly for uploading?

**Comment by muuvmuuv** [comment]
That is exactly what I discovered @wSedlacek I think the esbuild plugin should work with the Buffer instead of file output on filesystem in the end they should be the same. Or we need an Angular first-level support for Sentry (like with Tailwind), Astro also has announced this lately.

**Comment by lforst** [comment]
Interesting. Seems like Angular is doing unorthodox things here. I am not a fan of this. We might need to build something special for Angular. In the meanwhile you can use Sentry CLI to upload sourcemaps. That should still work.

**Comment by wSedlacek** [comment]
So I played around with the changeset #454 which used the `outputFiles` which does get the content of the files to the upload block. However, the files seem to be missing `sentry-dbid-` which prevents them from getting past this check.
https://github.com/wSedlacek/sentry-javascript-bundler-plugins/blob/a097e3249129d30a058d10f41a66960f17ca1647/packages/bundler-plugin-core/src/debug-id-upload.ts#L250-L256

Reverting the changes and looking at the outputs I am not seeing `sentry-dbid-` any of the files that are written to disk.
It seems like thus may be a multipart issue to get an esbuild plugin working with Angular.

**Comment by lforst** [comment]
@wSedlacek It's really cool that you tried to fix this. I had a suspicion that things are harder than they seem. Our biggest issue is frameworks doing shady stuff with bundlers.

**Comment by joewIST** [comment]
Also experiencing similar issues with sentry esbuild plugin, Angular and Nx combination. With some tweaking I can get rid of the errors, but mine just hangs on the "Building" step seemingly indefinitely...

**Comment by kevireilly** [comment]
Just wanted to provide the additional data point that I'm experiencing this issue with `@sentry/esbuild-plugin` and `splitting: true` specifically (without Angular). Does work appropriately for the main `esm` bundle.mjs and `iife` bundle.js.

**Comment by kevireilly** [comment]
By hacking around in `node_modules/@sentry/esbuild-plugin/dist/esm/index.mjs` I've gained some additional insight in to the issue and part of a potential solution. One problem is that only files of `kind === 'entry-point'` are being evaluated and looks like `dynamic-import` needs to be allowed as well.

https://github.com/getsentry/sentry-javascript-bundler-plugins/blob/main/packages/esbuild-plugin/src/index.ts#L54
```
if (args.kind !== "entry-point" && args.kind !== "dynamic-import") {
  return;
} else {
  ...various setup to later inject a Debug ID...
}
```

With this change, I'm observing named chunks getting uploaded appropriately. However, it does look like it is missing `chunk-[hash].[m]js` files presumably because they're not explicitly dynamically imported (at least in my scenario) as opposed to a byproduct of splitting that I don't understand just yet. At any rate, some good output where previously it was only just `esm-module.mjs.map` being identified and uploaded:
```
Source Map Upload Report
  Minified Scripts
    ~/0ae57e80-90ef-40a9-b1d0-5e522b79665a-4.js (sourcemap at middleware-XWW3I7LS.mjs.map, debug id 0ae57e80-90ef-40a9-b1d0-5e522b79665a)
    ~/49a9e9b6-6eda-47bb-9cd1-3a6dfc3427ea-3.js (sourcemap at esm-module.mjs.map, debug id 49a9e9b6-6eda-47bb-9cd1-3a6dfc3427ea)
    ~/5bec62e2-a022-4935-b999-9424fa2f28fc-0.js (sourcemap at goodies-2AHNUHGU.mjs.map, debug id 5bec62e2-a022-4935-b999-9424fa2f28fc)
  Source Maps
    ~/0ae57e80-90ef-40a9-b1d0-5e522b79665a-4.js.map (debug id 0ae57e80-90ef-40a9-b1d0-5e522b79665a)
    ~/49a9e9b6-6eda-47bb-9cd1-3a6dfc3427ea-3.js.map (debug id 49a9e9b6-6eda-47bb-9cd1-3a6dfc3427ea)
    ~/5bec62e2-a022-4935-b999-9424fa2f28fc-0.js.map (debug id 5bec62e2-a022-4935-b999-9424fa2f28fc)
```

**Comment by lforst** [comment]
@kevireilly Thanks for doing the research. I'll investigate whether also checking for `dynamic-import` could work. As it stands `splitting: true` is not supported.

**Comment by lforst** [comment]
@jahusa02 if there were you would see it here.

**Comment by attilacsanyi** [comment]
@lforst @kevireilly is there any update on this? We are using angular esbuild with `sourceMaps: true` which generates proper sourcemaps, but it seems [@sentry/esbuild-plugin](https://www.npmjs.com/package/@sentry/esbuild-plugin) not upload it. Thanks for any update on this issue.

**Comment by attilacsanyi** [comment]
@lforst what do you think, will be an update soon? We are using PWA and Angular with esbuild bundler and `@sentry/esbuild-plugin`. Until this issue the manual upload won't work as `sentry-cli` would update sources with `debug-id` and service worker will break.

_I want to avoid switching back to webpack as esbuild is more performant._

Grateful for any feedback. 🙇

**Comment by Lms24** [comment]
Hey @attilacsanyi! We're aware of the limitation around PWA/service workers and the CLI breaking the integrity of the generated artifacts. Sadly, as far as I'm aware, there's nothing we can do about that for the CLI flow as we'll never get around the timing problem here.

> but it seems [@sentry/esbuild-plugin](https://www.npmjs.com/package/@sentry/esbuild-plugin) not upload it

Just to confirm: Is the reason why our plugin doesn't upload the source maps the same that is discussed in this issue? 

Unfortunately, generally, there are no updates at the moment. We've all been very busy with the new SDK major version.

My 2 cents: Angular makes it extremely tricky for us to provide proper ways to upload source maps, due to them not allowing customized build setups. Having talked a bit about this with Angular team members recently, I'm afraid this won't change anytime soon :( I understand their reasoning but it sucks for tools like ours.

**Comment by muuvmuuv** [comment]
The main issue is still there because Angular does not create the files physically until build-ends so Sentry is not able to link them properly. As I state [a few comments earlier](https://github.com/getsentry/sentry-javascript-bundler-plugins/issues/351#issuecomment-1820889997) it could work if Sentry uses the Buffer during the build and hash as ID. But I am not deeply enough into that process, just did some digging.

**Comment by attilacsanyi** [comment]
> @attilacsanyi @Lms24 sorry for my ignorance, I might have forgotten. What is the problem with Sentry CLI + service workers again?

No problem @lforst . We use **Nx monorepo with Angular** and use `esbuild` [executor](https://nx.dev/nx-api/angular/executors/application#nxangularapplication).
Afaik after the build the bundle hash calculated and service worker script generated by Angular based on js bundles content.

The manual Sentry CLI script generated by `npx @sentry/wizard@latest -i sourcemaps` wizard I can manually link `inject`-ed `debug-ids` into sourcemaps and JS bundles.
```bash
"sentry:sourcemaps:inject": "sentry-cli sourcemaps inject --org MyOrg --project MyApp ./dist/apps/my-app/browser",
"sentry:sourcemaps:upload": "sentry-cli sourcemaps upload --org MyOrg --project MyApp ./dist/apps/my-app/browser --dist my-dist --release my-release",
"sentry:sourcemaps": "pnpm sentry:sourcemaps:inject && pnpm sentry:sourcemaps:upload"
```

But here is the problem the **bundles content has changed** as this debug ids were injected after Angular generated the hash for them hence **will be a hash mismatch**. _(The order of service worker hash generation and build order is managed by Angular builder so cannot hook into any process in the middle :( )_

Thank you if you have any feedback on this.

**Comment by lforst** [comment]
@attilacsanyi Ah thanks for explaining. Yeah this is a bit of a chicken and egg problem. I unfortunately don't see a clear path forward for us. In theory we are already operating and injecting debug IDs on the "virtual" level but it seems like angular is circumventing our plugin somehow.

In the meanwhile you can use the "legacy" (really should not be named as such because we cannot and will not remove it) way of uploading source maps: https://docs.sentry.io/platforms/javascript/sourcemaps/troubleshooting_js/legacy-uploading-methods/

**Comment by attilacsanyi** [comment]
> @attilacsanyi Ah thanks for explaining. Yeah this is a bit of a chicken and egg problem. I unfortunately don't see a clear path forward for us. In theory we are already operating and injecting debug IDs on the "virtual" level but it seems like angular is circumventing our plugin somehow.
> 
> In the meanwhile you can use the "legacy" (really should not be named as such because we cannot and will not remove it) way of uploading source maps: https://docs.sentry.io/platforms/javascript/sourcemaps/troubleshooting_js/legacy-uploading-methods/

Thanks for the idea @lforst, but based on the CI build logs the `uploadLegacySourcemaps` + `debug` approach has the same problem: it runs too early before Angular Esbuild finished generating the sourcemaps and not find anything to upload :(

```bash
...
# Build started here
2024-05-15T09:32:40.3550299Z > nx run my-app:build:production
2024-05-15T09:32:40.3550676Z 
2024-05-15T09:32:41.8097156Z - Building...
...
2024-05-15T09:33:28.3976701Z   chunk-PJQECVUI.js   139.3kb
2024-05-15T09:33:28.3977214Z   chunk-AHHMLF6Q.js   138.9kb
2024-05-15T09:33:28.3977626Z   chunk-X5XUYWHG.js   132.2kb
2024-05-15T09:33:28.3978152Z   ...and 358 more output files...
# Build finished here, but source maps not here yet
2024-05-15T09:33:28.3978615Z ⚡ Done in 46299ms
2024-05-15T09:33:29.7071376Z   INFO    2024-05-15 09:33:29.706584419 +00:00 Loaded config from /runner/_work/org/my_app/.sentryclirc
2024-05-15T09:33:29.7072988Z   DEBUG   2024-05-15 09:33:29.706628420 +00:00 sentry-cli version: 2.31.2, platform: "linux", architecture: "x86_64"
# Sentry CLI called to upload sourcemaps
2024-05-15T09:33:29.7076089Z   INFO    2024-05-15 09:33:29.706645140 +00:00 sentry-cli was invoked with the following command line: "/runner/_work/1inch-v2/1inch-v2/node_modules/@sentry/cli-linux-x64/bin/sentry-cli" "releases" "files" ....
...
# Sentry CLI did not find anything (too early execution)
2024-05-15T09:33:30.1550532Z > Nothing to upload
2024-05-15T09:33:31.1702563Z [sentry-esbuild-plugin] Warning: Didnt find any matching sources for debug ID upload. Please check the `sourcemaps.assets` option.
...
# Finally only here is the point when Angular Application bundle generation completes
2024-05-15T09:33:42.4374075Z  [37mApplication bundle generation complete. [60.584 seconds] [39m...
...
```
:(

**Comment by Lms24** [comment]
I'm realizing this is not ideal but you should probably get away with [using the CLI in the legacy way](https://docs.sentry.io/platforms/javascript/sourcemaps/troubleshooting_js/legacy-uploading-methods/#1-update-sdk-options). This you can call really after the build is finished (hence, the assets should be "physically" present) and the CLI won't touch your JS files to inject anything.

**Comment by pboling** [comment]
Not sure how helpful, but I'm getting the same error with latest SvelteKit (2.5.9) and none of the troubleshooting things are helping.  It's almost as if the config settings are not relevant to the error / behavior at all.
```
[sentry-vite-plugin] Warning: Didn't find any matching sources for debug ID upload. Please check the `sourcemaps.assets` option.
```

**Comment by pboling** [comment]
Let me know if I should create a new issue... finding some hints that all is not well in the library.

The [documentation](https://docs.sentry.io/platforms/javascript/guides/sveltekit/manual-setup/#configure-source-maps-upload) (typescript version) says we should be able to use `cleanArtifacts: true`.

![Screenshot 2024-05-19 at 10 51 02](https://github.com/getsentry/sentry-javascript-bundler-plugins/assets/19505/297b392c-e001-4f19-9703-f711fa4fdc68)

But it gives me an error:

![Screenshot 2024-05-19 at 10 54 09](https://github.com/getsentry/sentry-javascript-bundler-plugins/assets/19505/6c043aa0-79e3-42b9-a987-c055edf28505)

I'm on the latest version of all packages as of today.

**Comment by lforst** [comment]
@pboling Would you mind opening a separate issue for this? Thank you!

**Comment by Lms24** [comment]
Hey @pboling for the documentation part, I just merged a PR to the SvelteKit SDK docs that shows the updated `sentrySvelteKit` options. This should have happened a while ago but it unfortunately slipped through. Apologies for that. Also I opened another one, to show more options (https://github.com/getsentry/sentry-docs/pull/10094).

If you still experience issues after migrating to the new options, then please open a new issue, thanks :)

**Comment by brunoalano** [comment]
Any solution for this yet? Otherwise, did you opted to choose a different bundler than `esbuild`?

**Comment by attilacsanyi** [comment]
> Any solution for this yet? Otherwise, did you opted to choose a different bundler than `esbuild`?

@brunoalano I finally **kept the fast `esbuild` bundler** and [**do manual Sentry sourcemap upload**](https://docs.sentry.io/platforms/javascript/sourcemaps/uploading/cli/) after injecting debug ids.
I hope sometime really soon there will be a solution. I've [posted this issue to the Angular Community Discord](https://discord.com/channels/748677963142135818/1019773152269717555/threads/1250089600580321391) and also into [Nx Discord](https://discord.com/channels/1143497901675401286/1196825338831974420/1239951462080053321). 🤞

**Comment by lforst** [comment]
We are currently not working on this.

**Comment by joshdales** [comment]
👋 Hello I just ran into a similar issue with an Nx and Angular application when trying to inject debug ids after switching from webpack to esbuild to build.

The app is internationalized so have to use the sentry cli rather than the esbuild plugin in order inject them as otherwise it injects them before bundling the version for each language. 

When I run the cli command now I get the following error:
```
> Found 798 files
> Analyzing 798 sources
> Analyzing completed in 0.415s
> Injecting debug ids
error: Invalid embedded sourcemap in source file ./dist/prod/en-US/main-ICXOCCQC.js
```
But I don't get anyore information as to what has gone wrong and its only the main file that it has an issue with.

**Comment by lforst** [comment]
@joshdales Would you mind opening an issue over at https://github.com/getsentry/sentry-cli with reproduction details? Then we can take a look. Thanks!

**Comment by mcfedr** [comment]
@lforst  It would be nice if you could mention in the docs that the esbuild plugin doesnt work. I've wasted a bunch of time today trying to work out what I was doing wrong.

**Comment by lforst** [comment]
@mcfedr can you elaborate what exactly didn't work?

**Comment by mcfedr** [comment]
@lforst sorry, im probably being a bit rude. but what is described in this ticket. I spent the time to setup and install the plugin to my esbuild, and then spent quite a bit of time trying to work out why its not working, and then i realised its only uploading the entrypoint file, but none of the chunks that esbuild is generating. code splitting is a key part of what esbuild is doing in the build process for me. Now I'm trying to setup the sentry-cli version.

**Comment by Lms24** [comment]
@mcfedr so to confirm: Are you using esbuild with `splitting: true` configured?

**Comment by mcfedr** [comment]
> @mcfedr so to confirm: Are you using esbuild with `splitting: true` configured?

 Yes

**Comment by Lms24** [comment]
I opened https://github.com/getsentry/sentry-docs/pull/12205 to add a note to the esbuild source maps upload docs.

**Comment by lforst** [comment]
Closing because I don't think this is actionable any further.
